### DISCUSSION TITLE
How to extend a StateNotifier with non-model data

### DISCUSSION THREAD

**Discussion by scopendo** [rrousselGit/riverpod#959] [Q&A]
Hi,

I've got a situation where I have a model class `Category` that can be created on-device at runtime from a template category.  I have a corresponding state notifier class `CategoryController`.

I need to differentiate between whether the category exists on the device or not.  At first glance, this might be achieved by the controller holding an `AsyncValue<Category>`.  However, this doesn't allow me to differentiate between an existing or new category, as both would result in an `AsyncData` value.

An easy "watchable" way to understand whether the category exists would be to extend the `Category` class with a flag to indicate whether it exists, however, I would prefer to avoid that as that's merging a UI-runtime attribute with the real model attributes.

This leaves me with creating a wrapper class so that I end up with:

```dart
class ExtendedCategory {
  const ExtendedCategory(Category category, bool exists);
  final Category category;
  final bool exists;
}

class CategoryController extends StateNotifier<AsyncValue<ExtendedCategory>> {
  ...
  state = AsyncData(ExtendedCategory(category, exists));
  ...
}
```

I don't like this idea because then using the category values becomes a bit unnatural:

```dart
final categoryProvider = StateNotifierProvider<CategoryController, AsyncValue<ExtendedCategory>>(...);
...
final extendedCategory = ref.watch(categoryProvider);
print(extendedCategory.exists);
print(extendedCategory.category);
```

I think what I really want is a subclass of AsyncData, but I don't believe that is currently possible.

Wondering if anyone has a better approach.

**Comment by rrousselGit** [comment]
I'm a bit lost. What's the relationship between this "exists" and `AsyncData`?

`AsyncData` is purely for having a synchronous representation of a `Future`/`Stream`. It's unrelated to business logic

**Comment by scopendo** [comment]
The controller determines whether the category already exists as an async
operation.

Post-loading the notifier state should be either a category that already
exists or a category that does not already exist.

**Reply by rrousselGit** [reply]
Right, but why is that the responsibility of Asyncvalue and not category?

**Comment by scopendo** [comment]
Maybe it's not. I have been thinking that the category is an entity that
doesn't need to know whether it's been persisted, however, perhaps it
should do. Certainly things fall more neatly into place that way.

**Reply by scopendo** [reply]
To round off this topic, I'm addressing this design question by implementing an overarching "categories controller".  This also removes the issue.
